// @typedorm/common/esm/src/metadata/metadata-storage.js
export const IsAutoGenerateAttributeRawMetadataOptions = (attr) => !!attr.strategy;
export class MetadataStorage {
    _entities;
    _attributes;
    constructor() {
        this._entities = new Map();
        this._attributes = new Map();
    }
    /**
     * Get entity metadata by entity physical name
     * Physical name refers to value set to "name" property on @Entity
     * @param name entity physical name
     */
    getEntityByName(name) {
        return this.entities.find(en => en.name === name);
    }
    hasKnownEntity(entityClass) {
        return this._entities.has(entityClass);
    }
    getRawAttributesForEntity(entityClass) {
        const attributes = this._attributes.get(entityClass)?.values();
        if (!attributes) {
            // it is possible that entity might not have any attributes referenced such cases are
            // when inherited entity contains all the attribute declarations and derived ones only defines schema for it
            // Thus, instead of throwing an error, simply return [] list to continue processing other attributes
            return [];
        }
        return Array.from(attributes);
    }
    getRawEntityByTarget(entityClass) {
        const entity = this._entities.get(entityClass);
        if (!entity) {
            throw new Error(`No entity with name "${entityClass.name}" could be resolved, 
      make sure they have been declared at the connection creation time.`);
        }
        return entity;
    }
    addRawAttribute(entityClass, attribute) {
        let attributesForEntity = this._attributes.get(entityClass);
        if (!attributesForEntity) {
            attributesForEntity = new Map();
        }
        attributesForEntity.set(attribute.name, attribute);
        this._attributes.set(entityClass, attributesForEntity);
    }
    addRawEntity(entity) {
        this._entities.set(entity.target, entity);
    }
    get entities() {
        return Array.from(this._entities.values());
    }
    get attributes() {
        return Array.from(this._attributes.values()).map(attr => Array.from(attr.values()));
    }
}
