// @typedorm/core/esm/src/classes/connection/attribute-metadata-builder.js
import { ConflictingAttributeNameError, IsAutoGenerateAttributeRawMetadataOptions, MetadataManager, } from '@typedorm/common';
import { isUsedForPrimaryKey } from '../../helpers/is-used-for-primary-key';
import { AttributeMetadata } from '../metadata/attribute-metadata';
import { AutoGeneratedAttributeMetadata } from '../metadata/auto-generated-attribute-metadata';
export class AttributesMetadataBuilder {
    constructor() { }
    build(table, attributeTargetEntityClass, decoratedEntityClass) {
        const getRawAttributesForEntity = MetadataManager.metadataStorage.getRawAttributesForEntity(attributeTargetEntityClass);
        return getRawAttributesForEntity.map(attr => {
            const entityMetadata = this.tryGetEntityMetadata(attributeTargetEntityClass, decoratedEntityClass);
            // validate attribute names for primary key conflicts
            if ([
                entityMetadata.table?.partitionKey,
                entityMetadata.table?.usesCompositeKey() &&
                    entityMetadata.table?.sortKey,
            ]
                .filter(Boolean)
                .includes(attr.name)) {
                throw new ConflictingAttributeNameError({
                    attributeName: attr.name,
                    entity: entityMetadata.name,
                    tableName: entityMetadata.table?.name || '',
                });
            }
            if (IsAutoGenerateAttributeRawMetadataOptions(attr)) {
                return this.validateAndBuildAutoGeneratedAttributeMetadata(attr, entityMetadata);
            }
            return new AttributeMetadata({
                table,
                // when working with entity with multiple inheritance, use class with @Entity() if available
                entityClass: decoratedEntityClass ?? attributeTargetEntityClass,
                ...attr,
            });
        });
    }
    validateAndBuildAutoGeneratedAttributeMetadata(attr, entityMetadata) {
        if (attr.autoUpdate) {
            if (isUsedForPrimaryKey(entityMetadata.primaryKey, attr.name)) {
                throw new Error(`Failed to build metadata for "${attr.name}", attributes referenced in primary key cannot be auto updated.`);
            }
        }
        return new AutoGeneratedAttributeMetadata({
            ...attr,
        });
    }
    tryGetEntityMetadata(attributeTargetEntityClass, decoratedEntityClass) {
        let entityMetadata;
        try {
            // check if attribute target entity has metadata
            entityMetadata = MetadataManager.metadataStorage.getRawEntityByTarget(attributeTargetEntityClass);
        }
        catch (err) {
            // fall back to using derived class metadata
            entityMetadata =
                MetadataManager.metadataStorage.getRawEntityByTarget(decoratedEntityClass);
        }
        return entityMetadata;
    }
}
