// @typedorm/core/esm/src/classes/connection/connection-metadata-builder.js
import fg from 'fast-glob';
import path from 'path';
import { DuplicateEntityPhysicalNameError, MissingRequiredEntityPhysicalNameError, } from '@typedorm/common';
import { EntityMetadataBuilder } from './entity-metadata-builder';
import { MetadataManager } from '@typedorm/common';
export class ConnectionMetadataBuilder {
    connection;
    constructor(connection) {
        this.connection = connection;
    }
    buildEntityMetadatas(entities) {
        let possibleEntitiesToBuild = [];
        if (typeof entities === 'string') {
            possibleEntitiesToBuild = [...this.loadEntitiesFromDirs(entities)];
        }
        else {
            possibleEntitiesToBuild = [...entities];
        }
        // filter all entities that are not marked with `@Entity` decorator
        const entitiesToBuild = possibleEntitiesToBuild.filter(entity => MetadataManager.metadataStorage.hasKnownEntity(entity));
        const entityMetadatas = new EntityMetadataBuilder(this.connection).build(entitiesToBuild);
        this.sanitizeDuplicates(entityMetadatas);
        return entityMetadatas;
    }
    loadEntitiesFromDirs(pathMatchPattern) {
        const classesDirectory = path.normalize(pathMatchPattern);
        const allFiles = fg
            .sync(classesDirectory, {
            dot: false,
        })
            .map(file => require(file));
        return this.recursiveLoadModulesFromFiles(allFiles, []);
    }
    recursiveLoadModulesFromFiles(exported, allLoaded) {
        if (typeof exported === 'function') {
            allLoaded.push(exported);
        }
        else if (Array.isArray(exported)) {
            exported.forEach((file) => this.recursiveLoadModulesFromFiles(file, allLoaded));
        }
        else if (typeof exported === 'object' && exported !== null) {
            Object.keys(exported).forEach(key => this.recursiveLoadModulesFromFiles(exported[key], allLoaded));
        }
        return allLoaded;
    }
    sanitizeDuplicates(metadata) {
        const cache = [];
        metadata.forEach(metadata => {
            if (!metadata.name) {
                throw new MissingRequiredEntityPhysicalNameError(metadata.target.name);
            }
            if (cache.includes(metadata.name)) {
                throw new DuplicateEntityPhysicalNameError(metadata.name);
            }
            else {
                cache.push(metadata.name);
            }
        });
    }
}
