// @typedorm/common/esm/src/enums.js
export var RETURN_VALUES;
(function (RETURN_VALUES) {
    RETURN_VALUES["ALL_OLD"] = "ALL_OLD";
    RETURN_VALUES["ALL_NEW"] = "ALL_NEW";
    RETURN_VALUES["UPDATED_OLD"] = "UPDATED_OLD";
    RETURN_VALUES["UPDATED_NEW"] = "UPDATED_NEW";
    RETURN_VALUES["NONE"] = "NONE";
})(RETURN_VALUES || (RETURN_VALUES = {}));
export var INDEX_TYPE;
(function (INDEX_TYPE) {
    INDEX_TYPE["GSI"] = "GLOBAL_SECONDARY_INDEX";
    INDEX_TYPE["LSI"] = "LOCAL_SECONDARY_INDEX";
})(INDEX_TYPE || (INDEX_TYPE = {}));
export var QUERY_ORDER;
(function (QUERY_ORDER) {
    QUERY_ORDER["ASC"] = "ASC";
    QUERY_ORDER["DESC"] = "DESC";
})(QUERY_ORDER || (QUERY_ORDER = {}));
export var AUTO_GENERATE_ATTRIBUTE_STRATEGY;
(function (AUTO_GENERATE_ATTRIBUTE_STRATEGY) {
    /**
     * @summary Universally Unique Identifier version 4
     */
    AUTO_GENERATE_ATTRIBUTE_STRATEGY["UUID4"] = "UUID4";
    /**
     * @summary K-Sortable Unique Identifier
     */
    AUTO_GENERATE_ATTRIBUTE_STRATEGY["KSUID"] = "KSUID";
    AUTO_GENERATE_ATTRIBUTE_STRATEGY["EPOCH_DATE"] = "EPOCH_DATE";
    AUTO_GENERATE_ATTRIBUTE_STRATEGY["ISO_DATE"] = "ISO_DATE";
})(AUTO_GENERATE_ATTRIBUTE_STRATEGY || (AUTO_GENERATE_ATTRIBUTE_STRATEGY = {}));
export var INTERNAL_ENTITY_ATTRIBUTE;
(function (INTERNAL_ENTITY_ATTRIBUTE) {
    INTERNAL_ENTITY_ATTRIBUTE["ENTITY_NAME"] = "__en";
})(INTERNAL_ENTITY_ATTRIBUTE || (INTERNAL_ENTITY_ATTRIBUTE = {}));
export var CONSUMED_CAPACITY_TYPE;
(function (CONSUMED_CAPACITY_TYPE) {
    CONSUMED_CAPACITY_TYPE["INDEXES"] = "INDEXES";
    CONSUMED_CAPACITY_TYPE["TOTAL"] = "TOTAL";
    CONSUMED_CAPACITY_TYPE["NONE"] = "NONE";
})(CONSUMED_CAPACITY_TYPE || (CONSUMED_CAPACITY_TYPE = {}));
export var QUERY_SELECT_TYPE;
(function (QUERY_SELECT_TYPE) {
    QUERY_SELECT_TYPE["ALL_ATTRIBUTES"] = "ALL_ATTRIBUTES";
    QUERY_SELECT_TYPE["ALL_PROJECTED_ATTRIBUTES"] = "ALL_PROJECTED_ATTRIBUTES";
    QUERY_SELECT_TYPE["SPECIFIC_ATTRIBUTES"] = "SPECIFIC_ATTRIBUTES";
    QUERY_SELECT_TYPE["COUNT"] = "COUNT";
})(QUERY_SELECT_TYPE || (QUERY_SELECT_TYPE = {}));
