// @typedorm/core/cjs/src/classes/metadata/attribute-metadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AttributeMetadata = void 0;
const is_scalar_type_1 = require("./../../helpers/is-scalar-type");
const common_1 = require("@typedorm/common");
const build_primary_key_schema_1 = require("../../helpers/build-primary-key-schema");
const base_attribute_metadata_1 = require("./base-attribute-metadata");
class AttributeMetadata extends base_attribute_metadata_1.BaseAttributeMetadata {
    constructor(options) {
        const { name, entityClass, unique, table } = options;
        super(options);
        this.entityClass = entityClass;
        this.table = table;
        this.default = this.getDefaultValueProvider(name, options.default);
        if (unique) {
            this.unique = this.buildUniqueAttributesPrimaryKey(unique);
        }
    }
    getDefaultValueProvider(attrName, defaultValue) {
        if (!defaultValue) {
            return;
        }
        if ((0, is_scalar_type_1.isScalarTypeProvider)(defaultValue)) {
            return defaultValue;
        }
        else {
            return () => defaultValue;
        }
    }
    buildUniqueAttributesPrimaryKey(unique) {
        if ((0, common_1.IsPrimaryKey)(unique)) {
            return (0, build_primary_key_schema_1.buildPrimaryKeySchema)({
                table: this.table,
                primaryKey: unique,
                attributes: {
                    [this.name]: this.type,
                },
            });
        }
        else {
            return this.autoGeneratedPrimaryKeySchema();
        }
    }
    autoGeneratedPrimaryKeySchema() {
        const primaryKey = {};
        const uniqueKeyValue = `${common_1.DYNAMO_ATTRIBUTE_PREFIX}_${this.entityClass.name.toUpperCase()}.${this.name.toUpperCase()}#{{${this.name}}}`;
        if (this.table.usesCompositeKey()) {
            primaryKey.partitionKey = uniqueKeyValue;
            primaryKey.sortKey = uniqueKeyValue;
        }
        else {
            primaryKey.partitionKey = uniqueKeyValue;
        }
        return (0, build_primary_key_schema_1.buildPrimaryKeySchema)({
            table: this.table,
            primaryKey,
            attributes: {
                [this.name]: this.type,
            },
        });
    }
}
exports.AttributeMetadata = AttributeMetadata;
