// @typedorm/core/cjs/src/classes/expression/update/update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Update = void 0;
const base_expression_input_1 = require("../base-expression-input");
class Update extends base_expression_input_1.BaseExpressionInput {
    constructor() {
        super(...arguments);
        // empty prefix for base update type
        this.prefix = '';
    }
    getExpNameKey(key) {
        return `#UE_${key}`;
    }
    getExpValueKey(key) {
        return `:UE_${key}`;
    }
    /**
     * @override and
     */
    and() {
        this.expression += ', ';
        return this;
    }
    /**
     * @override or
     */
    or() {
        throw new Error('"or" operation is not supported with update expression.');
    }
    /**
     * Support merging multiple update expressions with same keyword
     * @override merge
     */
    merge(update) {
        const { expression, names, values, prefix } = update;
        if (!prefix) {
            throw new Error('Can not merge with Base `Update` type, merging expression must have a valid prefix.');
        }
        // if merging condition does not have anything to merge return
        if (!expression) {
            return this;
        }
        if (update.constructor === this.constructor) {
            this.expression += this.expression ? ',' : '';
            this.appendToExpression(expression);
        }
        else {
            this.appendToExpression(`${prefix} ${expression}`);
        }
        this.names = Object.assign(Object.assign({}, this.names), names);
        this.values = Object.assign(Object.assign({}, this.values), values);
        return this;
    }
    /**
     * Support merging multiple update expressions with same or different keywords
     * @override  mergeMany
     */
    mergeMany(inputs) {
        const merged = inputs.reduce((acc, upd) => {
            acc.merge(upd);
            return acc;
        }, this);
        return merged;
    }
    getSymbolForArithmeticOperator(operator) {
        const symbolMap = {
            INCREMENT_BY: '+',
            DECREMENT_BY: '-',
        };
        return symbolMap[operator];
    }
}
exports.Update = Update;
