// @typedorm/common/cjs/src/logger/debug-logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DebugLogger = exports.STATS_TYPE = exports.TRANSFORM_BATCH_TYPE = exports.TRANSFORM_SCAN_TYPE = exports.TRANSFORM_TRANSACTION_TYPE = exports.MANAGER_NAME = exports.TRANSFORM_TYPE = void 0;
const debug_1 = __importDefault(require("debug"));
const chalk_1 = __importDefault(require("chalk"));
const uuid_1 = require("uuid");
var TRANSFORM_TYPE;
(function (TRANSFORM_TYPE) {
    TRANSFORM_TYPE["GET"] = "GET";
    TRANSFORM_TYPE["PUT"] = "PUT";
    TRANSFORM_TYPE["UPDATE"] = "UPDATE";
    TRANSFORM_TYPE["DELETE"] = "DELETE";
    TRANSFORM_TYPE["QUERY"] = "QUERY";
    TRANSFORM_TYPE["RESPONSE"] = "RESPONSE";
})(TRANSFORM_TYPE = exports.TRANSFORM_TYPE || (exports.TRANSFORM_TYPE = {}));
var MANAGER_NAME;
(function (MANAGER_NAME) {
    MANAGER_NAME["ENTITY_MANAGER"] = "ENTITY MANAGER";
    MANAGER_NAME["TRANSACTION_MANAGER"] = "TRANSACTION MANAGER";
    MANAGER_NAME["BATCH_MANAGER"] = "BATCH MANAGER";
    MANAGER_NAME["SCAN_MANAGER"] = "SCAN_MANAGER";
})(MANAGER_NAME = exports.MANAGER_NAME || (exports.MANAGER_NAME = {}));
var TRANSFORM_TRANSACTION_TYPE;
(function (TRANSFORM_TRANSACTION_TYPE) {
    TRANSFORM_TRANSACTION_TYPE["TRANSACTION_WRITE"] = "TRANSACTION_WRITE";
    TRANSFORM_TRANSACTION_TYPE["TRANSACTION_READ"] = "TRANSACTION_READ";
})(TRANSFORM_TRANSACTION_TYPE = exports.TRANSFORM_TRANSACTION_TYPE || (exports.TRANSFORM_TRANSACTION_TYPE = {}));
var TRANSFORM_SCAN_TYPE;
(function (TRANSFORM_SCAN_TYPE) {
    TRANSFORM_SCAN_TYPE["SCAN"] = "SCAN";
    TRANSFORM_SCAN_TYPE["PARALLEL_SCAN"] = "PARALLEL_SCAN";
})(TRANSFORM_SCAN_TYPE = exports.TRANSFORM_SCAN_TYPE || (exports.TRANSFORM_SCAN_TYPE = {}));
var TRANSFORM_BATCH_TYPE;
(function (TRANSFORM_BATCH_TYPE) {
    TRANSFORM_BATCH_TYPE["BATCH_WRITE"] = "BATCH_WRITE";
    TRANSFORM_BATCH_TYPE["BATCH_READ"] = "BATCH_READ";
})(TRANSFORM_BATCH_TYPE = exports.TRANSFORM_BATCH_TYPE || (exports.TRANSFORM_BATCH_TYPE = {}));
var STATS_TYPE;
(function (STATS_TYPE) {
    STATS_TYPE["CONSUMED_CAPACITY"] = "CONSUMED_CAPACITY";
})(STATS_TYPE = exports.STATS_TYPE || (exports.STATS_TYPE = {}));
class DebugLogger {
    constructor() {
        // log
        this.debugTransformLog = (0, debug_1.default)('typedorm:transform:log');
        // batch transform logger
        this.debugTransformBatchLog = (0, debug_1.default)('typedorm:transform:batch:log');
        // transaction transform logger
        this.debugTransformTransactionLog = (0, debug_1.default)('typedorm:transform:transaction:log');
        // scan transform logger
        this.debugTransformScanLog = (0, debug_1.default)('typedorm:transform:scan:log');
        // info logger
        this.debugInfoLog = (0, debug_1.default)('typedorm:info:log');
        this.debugWarnLog = (0, debug_1.default)('typedorm:warn:log');
        this.debugErrorLog = (0, debug_1.default)('typedorm:error:log');
        // stats logger
        this.debugStatsLog = (0, debug_1.default)('typedorm:stats:log');
    }
    /**
     * Get unique request id for each request, and include it in each log
     * this allows for easy debugging
     * @returns unique request uuid
     */
    getRequestId() {
        return (0, uuid_1.v4)();
    }
    logTransform({ requestId, operation, prefix, entityName, primaryKey, body, options, }) {
        if (this.debugTransformLog.enabled) {
            this.debugTransformLog(`${chalk_1.default.bold.bgCyanBright(requestId)} ${chalk_1.default.green(operation)} ${chalk_1.default.blue(entityName)} ${chalk_1.default.magenta(prefix)}:`, ...(primaryKey
                ? [
                    chalk_1.default.blueBright('\nPrimary key: '),
                    chalk_1.default.white(this.ensurePrintable(primaryKey)),
                ]
                : []), ...(body
                ? [
                    chalk_1.default.blueBright('\nBody: '),
                    chalk_1.default.white(this.ensurePrintable(body)),
                ]
                : []), ...(options
                ? [
                    chalk_1.default.blueBright('\nOptions: '),
                    chalk_1.default.white(this.ensurePrintable(options)),
                ]
                : []));
        }
    }
    logTransformBatch({ requestId, operation, prefix, body, options, }) {
        if (this.debugTransformBatchLog.enabled) {
            this.debugTransformBatchLog(`${chalk_1.default.bold.bgCyanBright(requestId)} ${chalk_1.default.green(operation)} ${chalk_1.default.magenta(prefix)}:`, ...(body
                ? [
                    chalk_1.default.blueBright('\nBody: '),
                    chalk_1.default.white(this.ensurePrintable(body)),
                ]
                : []), ...(options
                ? [
                    chalk_1.default.blueBright('\nOptions: '),
                    chalk_1.default.white(this.ensurePrintable(options)),
                ]
                : []));
        }
    }
    logTransformTransaction({ requestId, operation, prefix, body, options, }) {
        if (this.debugTransformTransactionLog.enabled) {
            this.debugTransformTransactionLog(`${chalk_1.default.bold.bgCyanBright(requestId)} ${chalk_1.default.green(operation)} ${chalk_1.default.magenta(prefix)}:`, ...(body
                ? [
                    chalk_1.default.blueBright('\nBody: '),
                    chalk_1.default.white(this.ensurePrintable(body)),
                ]
                : []), ...(options
                ? [
                    chalk_1.default.blueBright('\nOptions: '),
                    chalk_1.default.white(this.ensurePrintable(options)),
                ]
                : []));
        }
    }
    logTransformScan({ requestId, prefix, operation, body, options, }) {
        if (this.debugTransformScanLog.enabled) {
            this.debugTransformScanLog(`${chalk_1.default.bold.bgCyanBright(requestId)} ${chalk_1.default.green(operation)} ${chalk_1.default.magenta(prefix)}:`, ...(body
                ? [
                    chalk_1.default.blueBright('\nBody: '),
                    chalk_1.default.white(this.ensurePrintable(body)),
                ]
                : []), ...(options
                ? [
                    chalk_1.default.blueBright('\nOptions: '),
                    chalk_1.default.white(this.ensurePrintable(options)),
                ]
                : []));
        }
    }
    logStats({ requestId, requestSegment, scope, statsType, consumedCapacityData, }) {
        if (this.debugStatsLog.enabled) {
            this.debugStatsLog(`${chalk_1.default.bold.bgCyanBright(requestId) +
                (requestSegment ? ':' + chalk_1.default.bold.bgCyanBright(requestId) : '')} ${chalk_1.default.green(scope)} ${chalk_1.default.magenta(statsType)}:`, chalk_1.default.white(this.ensurePrintable(consumedCapacityData)));
        }
    }
    logInfo({ requestId, scope, log, }) {
        if (this.debugInfoLog.enabled) {
            this.debugInfoLog(`${chalk_1.default.bold.bgCyanBright(requestId)} ${chalk_1.default.green(scope)}:`, chalk_1.default.white(this.ensurePrintable(log)));
        }
    }
    logWarn({ requestId, scope, log, }) {
        if (this.debugWarnLog.enabled) {
            this.debugWarnLog(`${chalk_1.default.bold.bgCyanBright(requestId)} ${chalk_1.default.green(scope)}:`, chalk_1.default.yellow(this.ensurePrintable(log)));
        }
    }
    logError({ requestId, scope, log, }) {
        if (this.debugErrorLog.enabled) {
            this.debugErrorLog(`${chalk_1.default.bold.bgCyanBright(requestId)} ${chalk_1.default.green(scope)}:`, chalk_1.default.red(this.ensurePrintable(log)));
        }
    }
    ensurePrintable(log) {
        if (typeof log === 'object' && log !== null) {
            return `\n${JSON.stringify(log, null, 2)}\n`;
        }
        else
            return log;
    }
}
exports.DebugLogger = DebugLogger;
