// @typedorm/core/cjs/public-api.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getScanManager = exports.getBatchManager = exports.getTransactionManger = exports.getEntityManager = exports.getConnection = exports.createConnections = exports.createConnection = exports.AutoGenerateAttributeValue = exports.Connection = void 0;
const container_1 = require("./src/classes/container");
const connection_manager_1 = require("./src/classes/connection/connection-manager");
// options
__exportStar(require("./src/classes/connection/connection-options"), exports);
// models
__exportStar(require("./src/classes/expression/condition"), exports);
__exportStar(require("./src/classes/expression/key-condition"), exports);
__exportStar(require("./src/classes/transaction/write-transaction"), exports);
__exportStar(require("./src/classes/transaction/read-transaction"), exports);
__exportStar(require("./src/classes/batch/write-batch"), exports);
__exportStar(require("./src/classes/batch/read-batch"), exports);
// managers
__exportStar(require("./src/classes/manager/entity-manager"), exports);
__exportStar(require("./src/classes/manager/batch-manager"), exports);
__exportStar(require("./src/classes/manager/transaction-manager"), exports);
__exportStar(require("./src/classes/manager/scan-manager"), exports);
// classes
var connection_1 = require("./src/classes/connection/connection");
Object.defineProperty(exports, "Connection", { enumerable: true, get: function () { return connection_1.Connection; } });
// helpers
var auto_generate_attribute_value_1 = require("./src/helpers/auto-generate-attribute-value");
Object.defineProperty(exports, "AutoGenerateAttributeValue", { enumerable: true, get: function () { return auto_generate_attribute_value_1.AutoGenerateAttributeValue; } });
// public method exports
function createConnection(options) {
    const connection = connectionManger().create(options);
    const connected = connection.connect();
    if (!connected) {
        throw new Error(`Failed to create connection with options: ${JSON.stringify(options)}`);
    }
    return connected;
}
exports.createConnection = createConnection;
function createConnections(optionsList) {
    return optionsList.map(options => createConnection(options));
}
exports.createConnections = createConnections;
function getConnection() {
    return container_1.Container.get(connection_manager_1.ConnectionManager).get();
}
exports.getConnection = getConnection;
function getEntityManager(connectionName) {
    return connectionManger().get(connectionName).entityManager;
}
exports.getEntityManager = getEntityManager;
function getTransactionManger(connectionName) {
    return connectionManger().get(connectionName).transactionManger;
}
exports.getTransactionManger = getTransactionManger;
function getBatchManager(connectionName) {
    return connectionManger().get(connectionName).batchManager;
}
exports.getBatchManager = getBatchManager;
function getScanManager(connectionName) {
    return connectionManger().get(connectionName).scanManager;
}
exports.getScanManager = getScanManager;
// private methods
function connectionManger() {
    return container_1.Container.get(connection_manager_1.ConnectionManager);
}
