// @typedorm/core/cjs/src/classes/transaction/write-transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WriteTransaction = void 0;
const transaction_1 = require("./transaction");
class WriteTransaction extends transaction_1.Transaction {
    constructor(
    /** only here for backwards compatibility
     * @deprecated
     * `WriteTransaction` does no longer need the connection object defined
     * at this level, it is now auto inferred by transaction transformer
     */
    connection, 
    /**
     * only here for backwards compatibility
     * @deprecated use `.add` for appending bulk items
     */
    transactionItems) {
        super();
        if (transactionItems && transactionItems.length) {
            throw new Error(`From 1.12.x, appending existing operation to 'WriteTransaction' is not supported. 
        Please use '.add' for appending bulk items.`);
        }
    }
    /**
     * @deprecated use operation specific method or `.add` instead
     */
    chian(chainedItem) {
        return this.add([chainedItem]);
    }
    add(transactionItems) {
        this.items.push(...transactionItems);
        return this;
    }
    addCreateItem(item, options) {
        this.items.push({
            create: {
                item,
                options: options,
            },
        });
        return this;
    }
    addUpdateItem(item, primaryKey, body, options) {
        this.items.push({
            update: {
                item,
                primaryKey,
                body,
                options: options,
            },
        });
        return this;
    }
    addDeleteItem(item, primaryKey, options) {
        this.items.push({
            delete: {
                item,
                primaryKey,
                options: options,
            },
        });
        return this;
    }
}
exports.WriteTransaction = WriteTransaction;
