// @typedorm/core/esm/src/classes/metadata/attribute-metadata.js
import { isScalarTypeProvider } from './../../helpers/is-scalar-type';
import { DYNAMO_ATTRIBUTE_PREFIX, IsPrimaryKey, } from '@typedorm/common';
import { buildPrimaryKeySchema } from '../../helpers/build-primary-key-schema';
import { BaseAttributeMetadata, } from './base-attribute-metadata';
export class AttributeMetadata extends BaseAttributeMetadata {
    unique;
    default;
    table;
    entityClass;
    constructor(options) {
        const { name, entityClass, unique, table } = options;
        super(options);
        this.entityClass = entityClass;
        this.table = table;
        this.default = this.getDefaultValueProvider(name, options.default);
        if (unique) {
            this.unique = this.buildUniqueAttributesPrimaryKey(unique);
        }
    }
    getDefaultValueProvider(attrName, defaultValue) {
        if (!defaultValue) {
            return;
        }
        if (isScalarTypeProvider(defaultValue)) {
            return defaultValue;
        }
        else {
            return () => defaultValue;
        }
    }
    buildUniqueAttributesPrimaryKey(unique) {
        if (IsPrimaryKey(unique)) {
            return buildPrimaryKeySchema({
                table: this.table,
                primaryKey: unique,
                attributes: {
                    [this.name]: this.type,
                },
            });
        }
        else {
            return this.autoGeneratedPrimaryKeySchema();
        }
    }
    autoGeneratedPrimaryKeySchema() {
        const primaryKey = {};
        const uniqueKeyValue = `${DYNAMO_ATTRIBUTE_PREFIX}_${this.entityClass.name.toUpperCase()}.${this.name.toUpperCase()}#{{${this.name}}}`;
        if (this.table.usesCompositeKey()) {
            primaryKey.partitionKey = uniqueKeyValue;
            primaryKey.sortKey = uniqueKeyValue;
        }
        else {
            primaryKey.partitionKey = uniqueKeyValue;
        }
        return buildPrimaryKeySchema({
            table: this.table,
            primaryKey,
            attributes: {
                [this.name]: this.type,
            },
        });
    }
}
