// @typedorm/core/cjs/src/classes/connection/attribute-metadata-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AttributesMetadataBuilder = void 0;
const common_1 = require("@typedorm/common");
const is_used_for_primary_key_1 = require("../../helpers/is-used-for-primary-key");
const attribute_metadata_1 = require("../metadata/attribute-metadata");
const auto_generated_attribute_metadata_1 = require("../metadata/auto-generated-attribute-metadata");
class AttributesMetadataBuilder {
    constructor() { }
    build(table, attributeTargetEntityClass, decoratedEntityClass) {
        const getRawAttributesForEntity = common_1.MetadataManager.metadataStorage.getRawAttributesForEntity(attributeTargetEntityClass);
        return getRawAttributesForEntity.map(attr => {
            var _a, _b, _c, _d;
            const entityMetadata = this.tryGetEntityMetadata(attributeTargetEntityClass, decoratedEntityClass);
            // validate attribute names for primary key conflicts
            if ([
                (_a = entityMetadata.table) === null || _a === void 0 ? void 0 : _a.partitionKey,
                ((_b = entityMetadata.table) === null || _b === void 0 ? void 0 : _b.usesCompositeKey()) &&
                    ((_c = entityMetadata.table) === null || _c === void 0 ? void 0 : _c.sortKey),
            ]
                .filter(Boolean)
                .includes(attr.name)) {
                throw new common_1.ConflictingAttributeNameError({
                    attributeName: attr.name,
                    entity: entityMetadata.name,
                    tableName: ((_d = entityMetadata.table) === null || _d === void 0 ? void 0 : _d.name) || '',
                });
            }
            if ((0, common_1.IsAutoGenerateAttributeRawMetadataOptions)(attr)) {
                return this.validateAndBuildAutoGeneratedAttributeMetadata(attr, entityMetadata);
            }
            return new attribute_metadata_1.AttributeMetadata(Object.assign({ table, 
                // when working with entity with multiple inheritance, use class with @Entity() if available
                entityClass: decoratedEntityClass !== null && decoratedEntityClass !== void 0 ? decoratedEntityClass : attributeTargetEntityClass }, attr));
        });
    }
    validateAndBuildAutoGeneratedAttributeMetadata(attr, entityMetadata) {
        if (attr.autoUpdate) {
            if ((0, is_used_for_primary_key_1.isUsedForPrimaryKey)(entityMetadata.primaryKey, attr.name)) {
                throw new Error(`Failed to build metadata for "${attr.name}", attributes referenced in primary key cannot be auto updated.`);
            }
        }
        return new auto_generated_attribute_metadata_1.AutoGeneratedAttributeMetadata(Object.assign({}, attr));
    }
    tryGetEntityMetadata(attributeTargetEntityClass, decoratedEntityClass) {
        let entityMetadata;
        try {
            // check if attribute target entity has metadata
            entityMetadata = common_1.MetadataManager.metadataStorage.getRawEntityByTarget(attributeTargetEntityClass);
        }
        catch (err) {
            // fall back to using derived class metadata
            entityMetadata =
                common_1.MetadataManager.metadataStorage.getRawEntityByTarget(decoratedEntityClass);
        }
        return entityMetadata;
    }
}
exports.AttributesMetadataBuilder = AttributesMetadataBuilder;
