// @typedorm/core/cjs/src/classes/connection/connection-metadata-builder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectionMetadataBuilder = void 0;
const fast_glob_1 = __importDefault(require("fast-glob"));
const path_1 = __importDefault(require("path"));
const common_1 = require("@typedorm/common");
const entity_metadata_builder_1 = require("./entity-metadata-builder");
const common_2 = require("@typedorm/common");
class ConnectionMetadataBuilder {
    constructor(connection) {
        this.connection = connection;
    }
    buildEntityMetadatas(entities) {
        let possibleEntitiesToBuild = [];
        if (typeof entities === 'string') {
            possibleEntitiesToBuild = [...this.loadEntitiesFromDirs(entities)];
        }
        else {
            possibleEntitiesToBuild = [...entities];
        }
        // filter all entities that are not marked with `@Entity` decorator
        const entitiesToBuild = possibleEntitiesToBuild.filter(entity => common_2.MetadataManager.metadataStorage.hasKnownEntity(entity));
        const entityMetadatas = new entity_metadata_builder_1.EntityMetadataBuilder(this.connection).build(entitiesToBuild);
        this.sanitizeDuplicates(entityMetadatas);
        return entityMetadatas;
    }
    loadEntitiesFromDirs(pathMatchPattern) {
        const classesDirectory = path_1.default.normalize(pathMatchPattern);
        const allFiles = fast_glob_1.default
            .sync(classesDirectory, {
            dot: false,
        })
            .map(file => require(file));
        return this.recursiveLoadModulesFromFiles(allFiles, []);
    }
    recursiveLoadModulesFromFiles(exported, allLoaded) {
        if (typeof exported === 'function') {
            allLoaded.push(exported);
        }
        else if (Array.isArray(exported)) {
            exported.forEach((file) => this.recursiveLoadModulesFromFiles(file, allLoaded));
        }
        else if (typeof exported === 'object' && exported !== null) {
            Object.keys(exported).forEach(key => this.recursiveLoadModulesFromFiles(exported[key], allLoaded));
        }
        return allLoaded;
    }
    sanitizeDuplicates(metadata) {
        const cache = [];
        metadata.forEach(metadata => {
            if (!metadata.name) {
                throw new common_1.MissingRequiredEntityPhysicalNameError(metadata.target.name);
            }
            if (cache.includes(metadata.name)) {
                throw new common_1.DuplicateEntityPhysicalNameError(metadata.name);
            }
            else {
                cache.push(metadata.name);
            }
        });
    }
}
exports.ConnectionMetadataBuilder = ConnectionMetadataBuilder;
