// @typedorm/core/cjs/src/helpers/auto-generate-attribute-value.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.autoGenerateValue = exports.AutoGenerateAttributeValue = void 0;
const ksuid_1 = __importDefault(require("ksuid"));
const common_1 = require("@typedorm/common");
const uuid_1 = require("uuid");
class AutoGenerateAttributeValue {
    static get UUID4() {
        return autoGenerateValue(common_1.AUTO_GENERATE_ATTRIBUTE_STRATEGY.UUID4);
    }
    static get KSUID() {
        return autoGenerateValue(common_1.AUTO_GENERATE_ATTRIBUTE_STRATEGY.KSUID);
    }
    static get ISO_DATE() {
        return autoGenerateValue(common_1.AUTO_GENERATE_ATTRIBUTE_STRATEGY.ISO_DATE);
    }
    static get EPOCH_DATE() {
        return autoGenerateValue(common_1.AUTO_GENERATE_ATTRIBUTE_STRATEGY.EPOCH_DATE);
    }
}
exports.AutoGenerateAttributeValue = AutoGenerateAttributeValue;
function autoGenerateValue(strategy) {
    switch (strategy) {
        case common_1.AUTO_GENERATE_ATTRIBUTE_STRATEGY.UUID4: {
            return (0, uuid_1.v4)();
        }
        case common_1.AUTO_GENERATE_ATTRIBUTE_STRATEGY.KSUID: {
            return ksuid_1.default.randomSync(new Date()).string;
        }
        case common_1.AUTO_GENERATE_ATTRIBUTE_STRATEGY.ISO_DATE: {
            return new Date().toISOString();
        }
        case common_1.AUTO_GENERATE_ATTRIBUTE_STRATEGY.EPOCH_DATE: {
            return Math.ceil(new Date().valueOf() / 1000);
        }
        default: {
            throw new common_1.UnknownAttributeValueGenerationStrategyError(strategy);
        }
    }
}
exports.autoGenerateValue = autoGenerateValue;
