// @typedorm/common/cjs/src/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_AWS_DYNAMO_DB_CLIENT_VERSION = exports.BATCH_READ_MAX_ALLOWED_ATTEMPTS = exports.BATCH_WRITE_MAX_ALLOWED_ATTEMPTS = exports.PARALLEL_SCAN_CONCURRENCY_LIMIT = exports.BATCH_WRITE_CONCURRENCY_LIMIT = exports.BATCH_READ_ITEMS_LIMIT = exports.BATCH_WRITE_ITEMS_LIMIT = exports.DYNAMO_QUERY_ITEMS_IMPLICIT_LIMIT = exports.DYNAMO_ATTRIBUTE_PREFIX = exports.TRANSACTION_READ_ITEMS_LIMIT = exports.TRANSACTION_WRITE_ITEMS_LIMIT = void 0;
/**
 * Current transaction write items limit set by dynamodb
 */
exports.TRANSACTION_WRITE_ITEMS_LIMIT = 100;
/**
 * Current transaction read items limit set by dynamodb
 */
exports.TRANSACTION_READ_ITEMS_LIMIT = 25;
/**
 * Prefix to apply to drm generated items
 */
exports.DYNAMO_ATTRIBUTE_PREFIX = 'DRM_GEN';
/**
 * Default Limit that is applied when query conditions do not include any explicit limits
 */
exports.DYNAMO_QUERY_ITEMS_IMPLICIT_LIMIT = 3000;
/**
 * Current batch write items limit set by dynamodb
 */
exports.BATCH_WRITE_ITEMS_LIMIT = 25;
/**
 * Current batch read items limit set by dynamodb
 */
exports.BATCH_READ_ITEMS_LIMIT = 100;
/**
 * Concurrency limit to apply, when running batch requests in parallel
 * i.e by default 5 promises are run at once
 */
exports.BATCH_WRITE_CONCURRENCY_LIMIT = 5;
/**
 * Concurrency limit to apply, when running parallel scan requests
 */
exports.PARALLEL_SCAN_CONCURRENCY_LIMIT = 10;
/**
 * Max number of allowed attempts for batch write items
 */
exports.BATCH_WRITE_MAX_ALLOWED_ATTEMPTS = 10;
/**
 * Max number of allowed attempts for batch write items
 */
exports.BATCH_READ_MAX_ALLOWED_ATTEMPTS = 10;
/**
 * Default DynamoDB client version to use
 */
exports.DEFAULT_AWS_DYNAMO_DB_CLIENT_VERSION = 2;
