// @typedorm/document-client/cjs/src/classes/document-client-v3.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocumentClientV3 = void 0;
const client_dynamodb_1 = require("@aws-sdk/client-dynamodb");
const lib_dynamodb_1 = require("@aws-sdk/lib-dynamodb");
const exceptions_1 = require("../exceptions");
const base_document_client_1 = require("./base-document-client");
class DocumentClientV3 extends base_document_client_1.DocumentClient {
    constructor(dynamoDBClient) {
        super();
        this.version = 3;
        const marshallOptions = {
            // Whether to automatically convert empty strings, blobs, and sets to `null`.
            convertEmptyValues: false,
            // Whether to remove undefined values while marshalling.
            removeUndefinedValues: false,
            // Whether to convert typeof object to map attribute.
            convertClassInstanceToMap: false, // false, by default.
        };
        const unmarshallOptions = {
            // Whether to return numbers as a string instead of converting them to native JavaScript numbers.
            wrapNumbers: false, // false, by default.
        };
        const translateConfig = { marshallOptions, unmarshallOptions };
        this.documentClient = lib_dynamodb_1.DynamoDBDocumentClient.from(dynamoDBClient, translateConfig);
    }
    put(input) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.documentClient.send(new lib_dynamodb_1.PutCommand(input));
        });
    }
    get(input) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.documentClient.send(new lib_dynamodb_1.GetCommand(input));
        });
    }
    query(input) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.documentClient.send(new lib_dynamodb_1.QueryCommand(input));
        });
    }
    update(input) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.documentClient.send(new lib_dynamodb_1.UpdateCommand(input));
        });
    }
    delete(input) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.documentClient.send(new lib_dynamodb_1.DeleteCommand(input));
        });
    }
    batchWrite(input) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.documentClient.send(new lib_dynamodb_1.BatchWriteCommand(input));
        });
    }
    batchGet(input) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.documentClient.send(new lib_dynamodb_1.BatchGetCommand(input));
        });
    }
    transactGet(input) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.documentClient.send(new lib_dynamodb_1.TransactGetCommand(input));
        });
    }
    transactWrite(input) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const response = yield this.documentClient.send(new lib_dynamodb_1.TransactWriteCommand(input));
                return response;
            }
            catch (err) {
                if (err instanceof client_dynamodb_1.TransactionCanceledException) {
                    // Remap TransactionCanceledException to unified TransactionCancelledException
                    throw new exceptions_1.TransactionCancelledException(err.Message || err.message, ((_a = err.CancellationReasons) === null || _a === void 0 ? void 0 : _a.map(reason => ({
                        code: reason.Code,
                        message: reason.Message,
                        item: reason.Item,
                    }))) || []);
                }
                throw err;
            }
        });
    }
    scan(input) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.documentClient.send(new lib_dynamodb_1.ScanCommand(input));
        });
    }
}
exports.DocumentClientV3 = DocumentClientV3;
